### PR TITLE
Fix Sweep Private keys

### PR DESCRIPTION
**Pull request by samholmes**
---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1201440626414086

### CODE REVIEW

File: src/common/utxobased/engine/makeUtxoEngineState.ts [hunk 1]
```diff
@@ -110,15 +110,21 @@ export function makeUtxoEngineState(
     updateTransactionCache: {}
   }
 
+  const clearObject = (obj: { [key: string]: any }): void => {
+    Object.keys(obj).forEach(key => {
+      // eslint-disable-next-line @typescript-eslint/no-dynamic-delete
+      delete obj[key]
+    })
+  }
```
Comment (thehobbit85) [inline src/common/utxobased/engine/makeUtxoEngineState.ts:120]: Why is this needed?
Doing something like this in Javascript is almost never needed and if it does, it usually means you are hiding some other issue.

File: src/common/utxobased/engine/makeUtxoEngine.ts [hunk 2]
```diff
@@ -492,28 +494,42 @@ export async function makeUtxoEngine(
...
+      ) => void
+      let failure: (err?: any) => void
+      const end = new Promise<EdgeTransaction>((resolve, reject) => {
+        success = resolve
+        failure = reject
+      })
+
+      // The temporary wallet info should include all formats for the currency
+      const allFormats = engineInfo.formats
+      if (allFormats == null) {
```
Comment (thehobbit85) [inline src/common/utxobased/engine/makeUtxoEngine.ts:520]: Why do you need to check for null? I thought "engineInfo.formats" should be strictly typed to exists no?

File: src/common/utxobased/engine/makeUtxoEngine.ts [hunk 3]
```diff
@@ -523,53 +539,67 @@ export async function makeUtxoEngine(
...
-
-          const utxos = (await processor.fetchUtxos({ utxoIds: [] })) as IUTXO[]
-          if (utxos === null || utxos.length < 1) {
-            throw new Error('Private key has no funds')
+          try {
+            // Delay to ensure that UTXOs are processed.
+            // This is a bug in the engine state; addresses are considered
+            // processed before the UTXOs are saved.
+            // TODO: Remove delay when fixed.
+            await delay(1000)
```
Comment (thehobbit85) [inline src/common/utxobased/engine/makeUtxoEngine.ts:565]: This should not even be as a temporary solution.
Lucky for us, the way to fix it is super simple, all you need to do is:

1. Add this `await args.onAddressChecked()` at line 1135 in 'makeUtxoEngineState.ts'
2. Change line 126 in  'makeUtxoEngineState.ts' to be: `processedCount += 0.5`

File: src/common/utxobased/engine/makeUtxoEngine.ts [hunk 4]
```diff
@@ -492,28 +494,42 @@ export async function makeUtxoEngine(
...
-      ) => void
-      let failure: (reason?: never) => void
-      const end: Promise<EdgeTransaction> = new Promise((resolve, reject) => {
-        success = resolve
-        failure = reject
-      })
-      const tmpDisklet = walletLocalDisklet
+
+      // Make temporary wallet disklet
+      const tmpDisklet = makeMemoryDisklet()
```
Comment (thehobbit85) [inline src/common/utxobased/engine/makeUtxoEngine.ts:542]: No problem here. I actually wanted to say that this is very clever and I really like it :)

File: src/common/utxobased/db/makeProcessor.ts [hunk 5]
```diff
@@ -260,6 +260,7 @@ export async function makeProcessor(
         // Return all UTXOs if no UTXO ids are specified
         if (utxoIds.length === 0) {
           const { data } = await tables.utxoById.dumpData('')
+          if (data[''] == null) return []
```
Comment (thehobbit85) [inline src/common/utxobased/db/makeProcessor.ts:263]: You can do this instead: `return Object.values(data[''] ?? {})`

File: src/common/utxobased/keymanager/keymanager.ts [hunk 6]
```diff
@@ -923,7 +923,7 @@ export async function makeTx(args: MakeTxArgs): Promise<MakeTxReturn> {
   const utxoPicker = coin.utxoPicker ?? utxopicker.utxoPicker
 
   const utxopicking =
-    args.forceUseUtxo != null ?? false
+    args.forceUseUtxo.length > 0
```
Comment (thehobbit85) [inline src/common/utxobased/keymanager/keymanager.ts:926]: You can always combine both tests into this: `(args.forceUseUtxo?.length ?? 0) > 0`

File: src/common/utxobased/engine/makeUtxoEngine.ts [hunk 7]
```diff
@@ -449,21 +449,32 @@ export async function makeUtxoEngine(
...
-
-          const address = await processor.fetchAddress(utxo.scriptPubkey)
-          if (address?.path == null) throw new Error('Invalid script pubkey')
-
-          return walletTools.getPrivateKey({ path: address.path, xprivKeys })
-        })
-      )
+      // Use the privateKeys (WIFs) from spendInfo, otherwise get them from the
+      // PBST inputs.
+      const privateKeys =
```
Comment (thehobbit85) [inline src/common/utxobased/engine/makeUtxoEngine.ts:454]: This should not be a ternary. Just use the normal `let/if/else` pattern.

**Review by thehobbit85 — CHANGES_REQUESTED** [review]

File: src/common/utxobased/engine/makeUtxoEngineState.ts (hunk 1 above)
Comment (samholmes) [reply src/common/utxobased/engine/makeUtxoEngineState.ts:120]: This is to avoid reference errors. It's a fix to the poor implementation. Each cache object should be a Map object (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/clear). Though I as I understand, Maps introduce challenges on some targeted platforms.

File: src/common/utxobased/engine/makeUtxoEngine.ts (hunk 2 above)
Comment (samholmes) [reply src/common/utxobased/engine/makeUtxoEngine.ts:520]: No. The current type is optional.

File: src/common/utxobased/engine/makeUtxoEngine.ts (hunk 3 above)
Comment (samholmes) [reply src/common/utxobased/engine/makeUtxoEngine.ts:565]: Clever solution. Though I will want to add a pretty verbose comment explaining why processedCount is only incremented by 0.5.

File: src/common/utxobased/engine/makeUtxoEngine.ts (hunk 4 above)
Comment (samholmes) [reply src/common/utxobased/engine/makeUtxoEngine.ts:542]: Thanks! 👍

File: src/common/utxobased/db/makeProcessor.ts (hunk 5 above)
Comment (samholmes) [reply src/common/utxobased/db/makeProcessor.ts:263]: Yes. Nice suggestion!

File: src/common/utxobased/keymanager/keymanager.ts (hunk 6 above)
Comment (samholmes) [reply src/common/utxobased/keymanager/keymanager.ts:926]: Beautiful!

File: src/common/utxobased/engine/makeUtxoEngine.ts (hunk 7 above)
Comment (samholmes) [reply src/common/utxobased/engine/makeUtxoEngine.ts:454]: I really wish we had `do` expressions. I dislike `let` for reasons. Can I use a IIFE with if/else?

Comment (thehobbit85) [reply src/common/utxobased/engine/makeUtxoEngine.ts:454]: > IIFE

Definitely

File: src/common/utxobased/keymanager/keymanager.ts (hunk 6 above)
Comment (samholmes) [reply src/common/utxobased/keymanager/keymanager.ts:926]: Wait, why do we need this? `forceUseUtxo` is never `undefined`.

**Review by thehobbit85 — APPROVED** [review]
